runtime/src/mm/cpp/Memory.cpp

```diff
@@ -125,23 +127,81 @@ extern "C" OBJ_GETTER(AllocArrayInstance, const TypeInfo* typeInfo, int32_t elem
-    // singleton will be created on a different thread and here we should check that, instead of creating
-    // another one (and registering `location` twice).
-    mm::GlobalsRegistry::Instance().RegisterStorageForGlobal(threadData, location);
-    TODO();
+
+    auto* object = mm::InitThreadLocalSingleton(threadData, location, typeInfo, ctor);
+    RETURN_OBJ(object);
+}
+
+extern "C" ALWAYS_INLINE OBJ_GETTER(InitSingleton, ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*)) {
+    auto* threadData = mm::ThreadRegistry::Instance().CurrentThreadData();
+
+    RETURN_RESULT_OF(mm::InitSingleton, threadData, location, typeInfo, ctor);
 }
 
 extern "C" RUNTIME_NOTHROW void InitAndRegisterGlobal(ObjHeader** location, const ObjHeader* initialValue) {
     auto* threadData = mm::ThreadRegistry::Instance().CurrentThreadData();
     mm::GlobalsRegistry::Instance().RegisterStorageForGlobal(threadData, location);
-    TODO();
+    mm::SetHeapRef(location, const_cast<ObjHeader*>(initialValue));
 }
 
 extern "C" const MemoryModel CurrentMemoryModel = MemoryModel::kExperimental;
 
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void SetStackRef(ObjHeader** location, const ObjHeader* object) {
+    mm::SetStackRef(location, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void SetHeapRef(ObjHeader** location, const ObjHeader* object) {
+    mm::SetHeapRef(location, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void ZeroHeapRef(ObjHeader** location) {
+    mm::SetHeapRef(location, nullptr);
+}
+
+extern "C" RUNTIME_NOTHROW void ZeroArrayRefs(ArrayHeader* array) {
+    for (uint32_t index = 0; index < array->count_; ++index) {
+        ObjHeader** location = ArrayAddressOfElementAt(array, index);
+        mm::SetHeapRef(location, nullptr);
+    }
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void ZeroStackRef(ObjHeader** location) {
+    mm::SetStackRef(location, nullptr);
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void UpdateStackRef(ObjHeader** location, const ObjHeader* object) {
+    mm::SetStackRef(location, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void UpdateHeapRef(ObjHeader** location, const ObjHeader* object) {
+    mm::SetHeapRef(location, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void UpdateHeapRefIfNull(ObjHeader** location, const ObjHeader* object) {
+    mm::CompareAndSwapHeapRef(location, nullptr, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW void UpdateReturnRef(ObjHeader** returnSlot, const ObjHeader* object) {
+    mm::SetStackRef(returnSlot, const_cast<ObjHeader*>(object));
+}
+
+extern "C" ALWAYS_INLINE RUNTIME_NOTHROW OBJ_GETTER(
+        SwapHeapRefLocked, ObjHeader** location, ObjHeader* expectedValue, ObjHeader* newValue, int32_t* spinlock, int32_t* cookie) {
+    auto* oldValue = mm::CompareAndSwapHeapRef(location, expectedValue, newValue);
```

Comment:
It is possible that at this moment `oldValue` is not referred from roots or heap. So this implementation is safe-points-specific.

Reply:
(Not sure if we should fix this, but maybe mark such places somehow).

Reply:
Do you mean it's specific to a particular safe point implementation, or to safe points in general?

Reply:
I mean that it doesn't expect GC to happen at this particular place, so it is incompatible with GCs that can stop threads at arbitrary points.

Reply:
Refactored a bit. Concerns about thread stopping mechanism should be limited to `ObjectOps` now. Also extracted initialization schemes into a different file, and they should now be oblivious to thread stopping technicalities.